### DISCUSSION TITLE
embedding servo in a C application

### DISCUSSION THREAD

**Discussion by vtorri** [servo/servo#32618] [Q&A] [answered]
I want to display epub content in a viewer application which is written in C. Are there some examples which show how to use servo in a C application ?

**Comment by mrobinson** [comment, marked as answer]
We do not have any examples of using Servo in a C application. You'd need to expose the entire Rust API via a C-compatible interface, which would likely be a good deal of work. We hope to have a C or C++ compatible API at some point though.

**Reply by vtorri** [reply]
any improvements about this since last year ?

**Reply by jdm** [reply]
We still have not exposed a C API for the Rust embedding API; it has gone through a lot of changes in the past 6 months.

**Comment by RolandHughes** [comment]
Do you have a link to real documentation? (I only just started looking into Servo but didn't find much on the main Web site.) Here's the situation:

Been in IT roughly 40 years. Started out on real computers that had to have real computer rooms, down to desktops, and finally spent the last decade in the embedded systems world, mostly medical devices. The tools in the medical devic/embedded systems world either A) suck   or   B) cost 27 vital organs and your first 6 children.

I had hoped that CopperSpice's port of Qt 4.8.x would amount to something, but they are too enamored with C++ compiler standards to create something usable. After years of trying to explain to them they will die poor and lonely going down that road, I finally had enough and recently [fully forked](https://github.com/RolandHughes/ls-cs) Project will get a rename and most likely move to SourceForge once it is further down [the roadmap](https://lscs-software.com/LsCs-Roadmap.html).

During the code cleanup and packaging stage it is becoming painfully obvious they have the old Webkit from 4.8.x days and just hacked things off to where it would "run." That has me poking around and after first looking at RDK (way too big of a whale I think) and absolutely refusing to use Blink due to Google's fetish for spy-ware, I stumbled on Servo. At first blush this looks like a good thing to put a C++ wrapper around so it can be a widget in an application. 

Don't need an answer immediately. Need to get all of the graphics back ends ripped out and replaced with only GLFW. (A large number of Qt bugs are the result of them rolling their own plugins for display instead of using a single robust well tested layer.) That's the point in time I think the project will get its rename. Post rename would be a good time to jettison the Webkit and start on a wrapper for the embedded browser widget.

Mostly thinking out loud here.

Oh. Until I'm done with the formal packaging scripts, I'm not merging into main. If you are looking for what is "now" you have to look at ls-cs-0.1.0 branch.

At any rate, a link to the documentation you have would be great!

**Reply by jdm** [reply]
https://doc.servo.org/servo/index.html
https://github.com/servo/servo/blob/main/components/servo/examples/winit_minimal.rs

**Comment by aidv** [comment]
I too would love to see a C or C++API.

I can't wait to replace WebVIew2 and WKWebView.

**Comment by mikkeljohnsen** [comment]
I have made wrapping for Gtk3/4 here: https://github.com/GSharpKit/servo-gtk/tree/servo_c_gtk

**Comment by mikkeljohnsen** [comment]
@jschwe You metioned the C-API for embedding. But I do not see any event for resizing / scrolling ?

Also is there a way to call js.Execute (script) ?

Will there be any direct access to the DOM, so you can manipulate it directly, without going through JS ?

**Reply by jschwe** [reply]
@mukilan would probably be able to tell you more, he's the main person who has worked on this.

**Reply by jdm** [reply]
>Will there be any direct access to the DOM, so you can manipulate it directly, without going through JS ?

There are no plans to allow this yet. It is also not available through the Rust embedding API.

**Reply by mukilan** [reply]
> You metioned the C-API for embedding. But I do not see any event for resizing / scrolling ?

Yes, not all WebView APIs are currently exposed in C. The focus of the work so far was to add the infrastructure for building and testing the C API as a shared library, so only the APIs that are needed for a basic integration test have been added. We plan to expand the API in the future. If the missing API for scrolling and resizing is blocking you and you are able to create a PR, I'd be happy to review it.

**Reply by mikkeljohnsen** [reply]
I currently have this: https://github.com/GSharpKit/servo-gtk/blob/servo_c_gtk/servo_c_gtk/servo/servo-webview.rs

Now it also exposes 'servo_webview_evaluate_script'

**Reply by RolandHughes** [reply]
Very cool!

**Comment by mikkeljohnsen** [comment]
@mukilan What I miss the most is printing or atleast print to PDF ...

Hope that will be possible soon.

**Reply by mukilan** [reply]
I think that would be orthogonal to the C API itself, as servo doesn't support printing currently. See https://github.com/servo/servo/issues/3788

**Comment by mg0269853-glitch** [comment]
#47092
